assignment5/assignment5.cpp

```diff
@@ -1,60 +1,103 @@
+	void DictionaryToGraph(){
+		graph.clear();
+		for(int i=0; i<dict.size()-1; i++){
+			for(int j=0; j<dict[i].size(); j++){
 
-		alphabet.clear();
-		int before, after;
-
-		for(int i = 0; i < dictionary.size()-1; i++){
-			for(int j=0; j < dictionary[i].size(); j++){
-
-				// Comparison is done with next word, same character
-				if(j < dictionary[i+1].size() && 
-					dictionary[i].substr(0, j) == dictionary[i+1].substr(0, j)){
-
-					if(dictionary[i][j]!=dictionary[i+1][j]){
-						before = alphabet.find(dictionary[i][j]);
-						after = alphabet.find(dictionary[i+1][j]);
-
-						if(before!= std::string::npos || after!=std::string::npos){
-							if(before!=std::string::npos && after == std::string::npos){
-								// after is not in alphabet yet
-								alphabet.insert(alphabet.begin() + before+1, dictionary[i+1][j]);
-							} else if( after != std::string::npos && before == std::string::npos){
-								// before is not in alphabet yet
-								alphabet.insert(alphabet.begin() + after, dictionary[i][j]);
-							} else if(after < before){
-								// both are, but the order is incorrect
-								alphabet.erase(before, 1);
-								alphabet.insert(alphabet.begin() + after, dictionary[i][j]);
-							}
-						} else {
-							// both characters are new to the alphabet
-							alphabet.push_back(dictionary[i][j]);
-							alphabet.push_back(dictionary[i+1][j]);
-						}
+				if(j < dict[i+1].size() && 
+					dict[i].substr(0, j) == dict[i+1].substr(0, j)){
+
+					if(dict[i][j]!=dict[i+1][j]){
+						graph[dict[i][j]].out_edges.push_back(dict[i+1][j]);
+						graph[dict[i+1][j]].in_edges.push_back(dict[i][j]);
 					}
 				} else {
 					break;
 				}
-
 			}
 
 		}
-		return alphabet;
 	}
-
+public:
+	UnknownLanguage(const std::vector<std::string> &dictionary): dict(dictionary){
+		DictionaryToGraph();
+	}
+	void SetDictionary(const std::vector<std::string> &dictionary){
+		dict = dictionary;
+	}
 	
+	std::string CalculateAlphabet(){
+		alphabet.clear();
+		if(graph.empty()) DictionaryToGraph();
+
+		std::string no_in_edges;
+		
+		for(std::map<char,Node>::iterator it=graph.begin(); it!=graph.end(); it++){
+			if (it->second.in_edges.empty()) no_in_edges.push_back(it->first);
+		}
+
+		char letter, child;
+		int position;
+		while(!no_in_edges.empty()){
+			letter = no_in_edges.back();
+			alphabet.push_back(letter);
+			no_in_edges.pop_back();
+			for(int i=0; i<graph[letter].out_edges.size(); i++){
+				child = graph[letter].out_edges[i];
+				position = graph[child].in_edges.find(letter);
+				graph[child].in_edges.erase(position, 1);
+				if (graph[child].in_edges.empty()) no_in_edges.push_back(child);
+			}
+			graph[letter].out_edges.clear();
+		}
+		graph.clear();
+		return alphabet;
+
+	}
 };
 
 int main(){
-	UnknownLanguage ul(std::vector<std::string> ({"ART","DRTC","DRAC","RAT","CAT","CAR"}));
-	std::cout << ul.CalculateAlphabet()<< std::endl;
+	UnknownLanguage ul(std::vector<std::string> ({"AD","AC","E","C"}));
```

Comment:
Yay! Unit tests!! :)
I like that you test the conditions needed and not that the result is exactly something (that may break).